vhdl_lang/src/analysis/completion.rs

```diff
@@ -2,17 +2,122 @@ use crate::analysis::DesignRoot;
 use crate::ast::visitor::{Visitor, VisitorResult};
 use crate::ast::{
     AnyDesignUnit, AnyPrimaryUnit, AnySecondaryUnit, ComponentDeclaration, Declaration,
-    EntityDeclaration, InstantiationStatement, InterfaceDeclaration, MapAspect, PackageDeclaration,
-    SubprogramDeclaration, UnitKey,
+    EntityDeclaration, HasUnitId, InstantiationStatement, InterfaceDeclaration, MapAspect,
+    ObjectClass, PackageDeclaration, SubprogramDeclaration, SubprogramDesignator, UnitKey,
+    WithDecl, WithRef,
 };
 use crate::data::{ContentReader, Symbol};
 use crate::syntax::Kind::*;
 use crate::syntax::{Symbols, Token, TokenAccess, Tokenizer, Value};
-use crate::{EntityId, Position, Source};
+use crate::{ast, EntityId, Position, Source};
+use fnv::FnvHashMap;
 use itertools::Itertools;
+use std::collections::hash_map::Entry;
 use std::collections::HashSet;
 use std::default::Default;
 
+#[derive(Debug, PartialEq, Clone)]
+pub enum CompletionKind {
+    Module,
+    Constant,
+    Signal,
+    Variable,
+    File,
+    Type,
+    Function,
+    Operator,
+    Attribute,
+    Alias,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum CompletionItemMode {
```

Comment:
This just seems like a duplicate of the lsp_types. Maybe the completion.rs file should not really be in the vhdl_lang crate but rather in the vhdl_ls crate and use the lsp_types directly. Completion is mostly a language server concern anyway.

Reply:
I generally agree. I'll see if it's simple to migrate.
The initial reason I put this into the `vhdl_lang` crate is because of the API design of `vhdl_ls`, which lacks some public features needed for completion. But maybe that has changed / is simpler than I initially thought.
Anyhow, it's probably a good idea to refactor this in the mid / long term (or at least before adding more completions and making the thing even more difficult to migrate).

Reply:
One thing that's especially annoying is that all the test modules are in `vhdl_lang` and they can't be used in `vhdl_ls` (please correct me if I have overlooked a very simple way of doing this). Especially the `Code` object and related methods would be nice to use.

Reply:
Technically crates can have test-code in the public API as well. But the fact that the test utilities also needs to be ported to move the functionaly is a cause for concern. Previously I tried to put the UI/presentation layer in vhdl_ls and the pure language knowledge in the vhdl_lang crate. Maybe you could just move the CompletionItemKind etc to vhdl_ls? They feel out of place inside vhdl_lang crate. vhdl_lang really only needs to give a list of `AnyEnt` that can be completed and the mapping from `AnyEnt` to CompletionItem type could happen in vhdl_ls.

Reply:
PS: So to summarize in the current architecture the vhdl_lang contains knowledge about the vhdl language and vhdl_ls uses it to create a language server. Thus I would avoid having copies of lsp_types in vhdl_lang itself since it is an indication that this architecture is not followed.

Reply:
I have done the refactoring. However, as you can see, the unit tests don't work because the `LibraryBuilder` and `Code` is not available outside of the `vhdl_lang` crate. What approach do you suggest for the `vhdl_ls` server? I can't seem to find a way to export `#[cfg(test)]` modules from `vhdl_lang`. Is there anything I have missed?

Reply:
I think you misunderstood what I meant by my previous comment. My point is that all the logic should be in `vhdl_lang` and the only thing that `vhdl_ls` does is to convert from `vhdl_lang` types into lsp-structures.  For example the code that creates all the completion options would be in vhdl_lang and tested in vhdl_lang. The only code that would be in vhdl_ls would be the code that converts an `AnyEnt` to a `CompletionItem`. The code that remains in `vhdl_ls` would then not be so important to test. 

For example you have `completion_item_from_decl` function. The code that is important to test is the one that generates the input to this function. The output of this function is just a conversion of a `vhdl_lang` internal type. 

If you look at how DocumentSymbol is implemented you will see that `vhdl_lang` tests all the logic needed to produce the document symbol and `vhdl_ls` is only the code that will convert the `vhdl_lang` internal representation into an lsp-type.

Reply:
Alright I think I got your idea now so you'll probably like the current version :D

Reply:
Yes I am happy with the new version. It follows the architecture I was trying to explain. Very nice!